OneCommunityGlobal/HGNRest
# Shengwei add cache for informations api

# Description
![image](https://github.com/OneCommunityGlobal/HGNRest/assets/69882989/15f08476-8ed6-42d3-a611-5ea63ae0e23c)

## Related PRS (if any):


## Main changes explained:
- Modify file src/controllers/informationController.js for adding cache handling in CRUD operation.


## How to test:
1. check into the current branch and the latest development branch in frontend
2. do `npm install` and `...` to run this PR locally
3. Clear site data/cache
4. log as admin/owner or any other roles 
5. To inspect: open browser -> right-click on page and click inspect -> go to inspect mode -> go to network
6. To test in Web UI: go to localhost:3000, the default page /dashboard will make multiple Information API calls 
   - **verify** that any call to `/api/informations` **has significant performance improvement after the initial call.** 
   - You could refresh the page and check the Time in the Network tab under the Inspect mode. (Should have around 700 ms end-to-end response time in the first call) 
   - Initial API call
![image](https://github.com/OneCommunityGlobal/HGNRest/assets/69882989/a9a03849-3af8-4705-957c-8c3005b94713)

   - Later API calls with the cache 
![image](https://github.com/OneCommunityGlobal/HGNRest/assets/69882989/cc97ae28-fa2a-47fd-9ef3-b244fa789169)
![image](https://github.com/OneCommunityGlobal/HGNRest/assets/69882989/43a5347f-ef21-4ffb-9ae9-91353bd09fc8)

7. To test with backend debug mode: Enable debug mode in VS code and add breakpoints in the code in `informationController.js` file.
![image](https://github.com/OneCommunityGlobal/HGNRest/assets/69882989/a95c6d6e-1047-4214-9f10-319129ed3c16)
     -  Verify the very first call should not hit the cache 
![image](https://github.com/OneCommunityGlobal/HGNRest/assets/69882989/553abd5c-7215-4c1a-a578-966bc566cbaa)

     -  Verify that any later call should hit the cache if no records have changed.
![image](https://github.com/OneCommunityGlobal/HGNRest/assets/69882989/490f086e-ddef-43b9-90ec-cd5d650406ad)

8. Unfortunately, the update operation for information records is not implemented in the frontend. Hence, user is not able to test cache removal in endpoint ‘**updateInformation**’, '**addInformation**' , and '**deleteInformation**' in the frontend UI.

## Screenshots or videos of changes:
**Start around 0:27**

https://github.com/OneCommunityGlobal/HGNRest/assets/69882989/0d5050ff-a530-4922-bbb9-e352d118bb22

## Note:
